### ISSUE TITLE
Pyright fails to validate TypedDict return type

### ISSUE THREAD

**Issue by diego-pm** [microsoft/pyright#3841]
**Describe the bug**
When building a function that expects a TypedDict as a return type, sometimes pyright emits a false positive indicating that the return type is not compatible.

**To Reproduce**
pyright version: 1.1.267
python version: 3.7

Execute pyright on the following script:

```python
from typing import List
from typing_extensions import TypedDict


class _ColorDict(TypedDict):
    id: int
    color: str


class _ColorsResponseDict(TypedDict):
    count: int
    colors: List[_ColorDict]


def function(colors: List[tuple]) -> _ColorsResponseDict:
    res = {
        "count": len(colors),
        "colors": [{"id": c[0], "color": c[1]} for c in colors],
    }
    return res
```

The error does not occur if inside the function I return directly the dictionary without assigning it to a variable first:

```python
def function(colors: List[tuple]) -> _ColorsResponseDict:
    return {
        "count": len(colors),
        "colors": [{"id": c[0], "color": c[1]} for c in colors],
    }
```

**Expected behavior**
No errors should be raised by pyright.

**VS Code extension or command-line**
Checked in both pyright CLI (1.1.267) and pylance

**Comment by erictraut** [comment]
This behavior is intended. In your first code sample, you are assigning a dictionary literal expression to a local variable named `res`. This local variable has no declared type, so the type of the expression is inferred using normal dictionary inference rules. See [this documentation](https://github.com/microsoft/pyright/blob/main/docs/type-inference.md#dictionary-expressions) for details. 

If you use the same dictionary expression in a context where there is an "expected type", a static type checker like pyright or mypy will employ a technique called [bidirectional inference](https://github.com/microsoft/pyright/blob/main/docs/type-inference.md#bidirectional-type-inference-expected-types). The "expected type" can come from a variable type declaration (if the expression is assigned to that variable), a parameter type declaration (if the expression is used as an argument in a call expression), or a return type declaration (if the expression is used in a return statement). This is why it works in your second example.

To fix your first code sample, you would need to provide a type annotation for the `res` local variable.

```python
def function(colors: List[tuple]) -> _ColorsResponseDict:
    res: _ColorsResponseDict = {
        "count": len(colors),
        "colors": [{"id": c[0], "color": c[1]} for c in colors],
    }
    return res
```
